File: tmdb-core/src/main/java/dev/reuss/tmdb/core/http/QueryParams.java
package dev.reuss.tmdb.core.http;

import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Mutable builder for TMDB query parameters.
 *
 * <p>This class is used to collect request query parameters before creating
 * a {@link TmdbRequest}. Parameters with {@code null} or blank values are
 * ignored, which makes it convenient to add optional request settings such
 * as language, region or page.</p>
 *
 * <p>The insertion order is preserved while building the parameters. The map
 * returned by {@link #toMap()} is an immutable copy.</p>
 */
public final class QueryParams {

    private final Map<String, String> values = new LinkedHashMap<>();

    private QueryParams() {
    }

    /**
     * Creates an empty query parameter builder.
     *
     * @return a new query parameter builder
     */
    public static QueryParams create() {
        return new QueryParams();
    }

    /**
     * Adds a query parameter if the given value is not {@code null} or blank.
     *
     * @param name  the query parameter name
     * @param value the query parameter value
     * @return this builder
     * @throws IllegalArgumentException if the parameter name is {@code null}, blank or empty
     */
    public QueryParams add(String name, String value) {
        if (name == null || name.isBlank()) {
            throw new IllegalArgumentException("Query parameter name must not be blank");
        }

        if (value != null && !value.isBlank()) {
            values.put(name, value);
        }

        return this;
    }

    /**
     * Adds a query parameter if the given value is not {@code null}.
     *
     * <p>The value is converted using {@link Object#toString()}.</p>
     *
     * @param name  the query parameter name
     * @param value the query parameter value
     * @return this builder
     * @throws IllegalArgumentException if the parameter name is {@code null}, blank or empty
     */
    public QueryParams add(String name, Object value) {
        if (value == null) {
            return this;
        }

        return add(name, value.toString());
    }

    /**
     * Returns the collected query parameters as an immutable map.
     *
     * @return an immutable copy of the collected query parameters
     */
    public Map<String, String> toMap() {
        return Map.copyOf(values);
    }
}

File: tmdb-core/src/main/kotlin/dev/reuss/tmdb/core/http/QueryParams.kt
package dev.reuss.tmdb.core.http

import java.util.Collections

/**
 * Insertion-ordered builder for TMDB query parameters.
 *
 * This class is used to collect request query parameters before creating
 * a [TmdbRequest]. Parameters with `null` or blank values are ignored, which
 * makes it convenient to add optional request settings such as language,
 * region or page.
 *
 * Adding the same name again replaces its value without changing its position.
 * [toMap] returns an immutable snapshot, so later additions do not affect an
 * already created map.
 */
class QueryParams private constructor() {
    private val values = linkedMapOf<String, String>()

    /**
     * Adds or replaces a parameter when [value] is neither `null` nor blank.
     *
     * @throws IllegalArgumentException if [name] is blank, even when [value] is absent
     */
    fun add(
        name: String,
        value: String?,
    ): QueryParams =
        apply {
            require(name.isNotBlank()) {
                "Query parameter name must not be blank"
            }

            if (!value.isNullOrBlank()) {
                values[name] = value
            }
        }

    /**
     * Adds or replaces a parameter when [value] is not `null`.
     *
     * The value is converted using [Any.toString].
     */
    fun add(
        name: String,
        value: Any?,
    ): QueryParams {
        if (value == null) {
            return this
        }

        return add(name, value.toString())
    }

    /**
     * Creates an immutable snapshot of the current parameters.
     */
    fun toMap(): Map<String, String> = Collections.unmodifiableMap(LinkedHashMap(values))

    companion object {
        /** Creates an empty parameter builder. */
        @JvmStatic
        fun create(): QueryParams = QueryParams()
    }
}
